### PR TITLE
Add Dotnet 8 Lambda runtime

### PR DESCRIPTION
**Pull request by yongliu-mdsol**
## Motivation
AWS released sam/dotnet8 image about a week ago

EDITED (@joe4dev):
AWS recently (2024-02-22) introduced the new Lambda runtime Dotnet8. We should support it in LocalStack.

- https://aws.amazon.com/blogs/compute/introducing-the-net-8-runtime-for-aws-lambda/
- https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

## Changes
Adding dotnet 8 runtime to localstack

EDITED  (@joe4dev):
* Re-structure the Dotnet build to improve maintenance by getting rid of runtime-specific values
* Add transformer for CodeSha256 because that changes on AWS => re-generate all Lambda common test snapshots
* Clean up unused Lambda packages

## Testing

see shadow LocalStack PR for ext testing https://github.com/localstack/localstack/pull/10504

## TODO

What's left to do:

- [x] Fix flaky multi-build (see comment)

### CODE REVIEW

**Comment by yongliu-mdsol** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by yongliu-mdsol** [comment]
recheck

**Comment by yongliu-mdsol** [comment]
hmm, I can't add labels to the pr

**Comment by yongliu-mdsol** [comment]
@dfangl @joe4dev @dominikschubert can you please help me with this PR? some checks fails due to missing git token?

**Comment by dfangl** [comment]
Hi @yongliu-mdsol !
The missing token is not a problem. However, there are some things left to do before we can merge this PR:
1. Please follow this instructions for adding a new runtime https://github.com/localstack/localstack/blob/master/localstack/services/lambda_/runtimes.py#L4-L28
2. Please do not commit any binary files to the repository. There is a complete `dotnet8` folder which does not seem to belong in there.
3. The snapshots have to be regenerated like instructed in the linked instructions. Also the order of the exception messages in that file is determined by the aws validated snapshots. More information can be found here: https://docs.localstack.cloud/contributing/integration-tests/#test-against-amazon-web-services 
4. The `localstack/aws/api/lambda_/__init__.py` file should not be manually edited, but needs to be auto generated as instructed in the instructions.

If you need any help with any of these steps, please do not hesitate to ask! If this is too much of a hassle for you, please also tell us and we will take over!

**Comment by yongliu-mdsol** [comment]
@dfangl Thanks for the comment. I've cleaned up the changes a bit. I'll check it out on the snapshot testing. It may take me some time. meanwhile, if you can fix the issue with snapshot, it will be great.

**Comment by yongliu-mdsol** [comment]
hi @dfangl, I was able to have integration tests passed locally. However, when I tried to regenerate snapshot for the test file itself, some of other runtime tests like nodejs. most of those complains about that aws no longer support those versions.
Would you please help me here? thanks

**Comment by joe4dev** [comment]
> hi @dfangl, I was able to have integration tests passed locally. However, when I tried to regenerate snapshot for the test file itself, some of other runtime tests like nodejs. most of those complains about that aws no longer support those versions. Would you please help me here? thanks

Thank you for reaching out @yongliu-mdsol 🙏 You are right 👍 . Especially with yesterday's new wave of [Lambda runtime deprecations](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-deprecated), we have 6 deprecated runtimes that need to be updated and tests need to be adjusted accordingly. These deprecations also unblock some long-awaited cleanups given that all supported Lambda runtimes are now ARM-compatible.

I started tackling these updates in a separate PR https://github.com/localstack/localstack/pull/10441
I will let you know once these changes are released and this PR is ready for rebasing. I can do the rebasing and snapshot updates if you prefer.

**Comment by yongliu-mdsol** [comment]
@joe4dev yes please. if you can help me with this pr, that will be great. we are waiting for dotnet 8 in localstack to upgrade our projects.

**Comment by joe4dev** [comment]
@yongliu-mdsol The separate PR with the deprecation updates is finally merged https://github.com/localstack/localstack/pull/10441 🙌 . It took a bit longer than expected because we took the chance to clean up our Lambda (pre-)build process and make it fully ARM-compatible. This included CI changes and further changes to our internal Pro version.

I rebased your PR and edited it to align the Dotnet builds with other runtimes. We wanted to make it easier, faster, and more maintainable to add new Dotnet runtimes in the future. Therefore, I force-pushed the rebase and subsequent updates here. I hope that's ok for you. We should be able to get this one in quickly now 🚀

File: tests/aws/services/lambda_/functions/common/README.md
Comment (joe4dev) [inline tests/aws/services/lambda_/functions/common/README.md:40]: * We waste one extra build cycle for the top-level `dotnet` directory.

File: tests/aws/services/lambda_/functions/common/echo/dotnet/Makefile
Comment (joe4dev) [inline tests/aws/services/lambda_/functions/common/echo/dotnet/Makefile:27]: There is currently a flaky issue when building two Dotnet runtimes in short succession:

```
cd localstack/tests/aws/services/lambda_/functions/common/echo
# FAILS
make -C dotnet build && make -C dotnet6 build
# WORKS
make -C dotnet build && sleep 1 && make -C dotnet6 build
```

Error:
```
Zipping publish folder /app2/bin/Release/net6.0/publish to /out/handler.zip
... zipping: zip I/O error: No such file or directory
... zipping: zip error: Could not create output file (/out/handler.zip)
Lambda project successfully packaged: /out/handler.zip
cp: build/handler.zip: No such file or directory
make[2]: *** [build] Error 1
make[1]: *** [build] Error 2
make: *** [dotnet6] Error 2
```

Maybe that's related to the Docker container, which needs some time to shut down 🤔 
@dfangl Do you have any suggestions on how we should guard against this to avoid a flaky build and fix it locally

Comment (dfangl) [reply tests/aws/services/lambda_/functions/common/echo/dotnet/Makefile:27]: Very possible that there is a race condition here. Maybe we could avoid having the same mount for both? Passing in the build folder so we have a different one for dotnet6 and dotnet8 builds?

Comment (joe4dev) [reply tests/aws/services/lambda_/functions/common/echo/dotnet/Makefile]: Good idea 👍 
I fixed it using the `FRAMEWORK` version as a natural build directory suffix to minimize sub-Makefile changes.

File: tests/aws/services/lambda_/test_lambda_common.py
Comment (dfangl) [inline tests/aws/services/lambda_/test_lambda_common.py:144]: So those are only set in LocalStack, but not in AWS?

**Review by dfangl — APPROVED** [review]
LGTM, nice work! Thank you @yongliu-mdsol for the PR and @joe4dev for bringing it over the line!

File: tests/aws/services/lambda_/test_lambda_common.py
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_common.py:144]: correct 👍 
(I can sneak in a clarifying comment in a later PR)

**Comment by yongliu-mdsol** [comment]
Thanks @joe4dev for fixing up my PR. 

@dfangl do you know if this will be part of next release? and when will that be? even a prerelease image would be sufficient for us.

**Comment by dfangl** [comment]
@yongliu-mdsol it will be part of the next release, which is scheduled for Thursday.

**Pull request merged**
